gmso/external/convert_parmed.py

```diff
@@ -339,146 +386,296 @@ def _angle_types_from_pmd(structure, angle_types_member_map=None):
+    if len(improper_dihedralsList) == 0 or not structure.dihedral_types:
+        return pmd_top_impropertypes
+    unique_improper_types = list(
+        dict.fromkeys(
+            [
+                (
+                    improper.type.phi_k,
+                    improper.type.phase,
+                    improper.type.per,
+                    (atomtype_namegetter(improper)),
+                )
+                for improper in improper_dihedralsList
+            ]
+        )
     )
-
-    for dihedraltype in structure.dihedral_types:
+    for impropertype in unique_improper_types:
         improper_params = {
-            "k": (dihedraltype.phi_k * u.Unit("kcal / mol")),
-            "phi_eq": (dihedraltype.phase * u.degree),
-            "n": dihedraltype.per * u.dimensionless,
+            "k": (impropertype[0] * u.Unit("kcal / mol")),
+            "phi_eq": (impropertype[1] * u.degree),
+            "n": impropertype[2] * u.dimensionless,
         }
-        expr = lib["PeriodicImproperPotential"]
-        member_types = improper_types_member_map.get(id(dihedraltype))
-        top_impropertype = gmso.ImproperType.from_template(
-            potential_template=expr, parameters=improper_params
+        periodic_torsion_potential = lib["PeriodicImproperPotential"]
+        name = periodic_torsion_potential.name
+        expression = periodic_torsion_potential.expression
+        variables = periodic_torsion_potential.independent_variables
+
+        top_impropertype = gmso.ImproperType(
+            name=name,
+            parameters=improper_params,
+            expression=expression,
+            independent_variables=variables,
+            member_types=impropertype[3],
         )
-        pmd_top_impropertypes[id(dihedraltype)] = top_impropertype
-        top_impropertype.member_types = member_types
+        pmd_top_impropertypes[impropertype] = top_impropertype
+
+    return pmd_top_impropertypes
+
+
+def _improper_types_harmonic_from_pmd(
+    structure,
+):
+    """Convert ParmEd improper types to GMSO improperType.
 
-    for impropertype in structure.improper_types:
+    This function take in a Parmed Structure, iterate through its
+    improper_types, create a corresponding
+    GMSO.improperType, and finally return a dictionary containing all
+    pairs of pmd.impropertype and GMSO.improperType
+    Parameters
+    ----------
+    structure: pmd.Structure
+        Parmed Structure that needed to be converted.
+
+    Returns
+    -------
+    pmd_top_impropertypes : dict
+        A dictionary linking a pmd.improperType
+        object to its corresponding GMSO.improperType object.
+    """
+    pmd_top_impropertypes = dict()
+    if not structure.improper_types:
+        return pmd_top_impropertypes
+    unique_improper_types = list(
+        dict.fromkeys(
+            [
+                (
+                    improper.type.psi_k,
+                    improper.type.psi_eq,
+                    (atomtype_namegetter(improper)),
```

Comment:
same comment as above regarding sorting

Reply:
Addressed in [408ef1b](https://github.com/mosdef-hub/gmso/pull/721/commits/408ef1ba5b6b613ad3f9f541d395a4a99daaa1e5)